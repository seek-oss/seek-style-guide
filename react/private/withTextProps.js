import React from 'react';
import PropTypes from 'prop-types';

import omit from 'lodash/omit';
import has from 'lodash/has';
import some from 'lodash/some';
import includes from 'lodash/includes';
import forEach from 'lodash/forEach';

export const sizes = [
  'small',
  'standard',
  'large',
  'subheading',
  'heading',
  'headline',
  'hero'
];

const getBooleanSizePropTypes = () => {
  const booleanProps = {};

  forEach(sizes, size => {
    booleanProps[size] = PropTypes.bool;
  });

  return booleanProps;
};

export const SizePropTypes = {
  // eslint-disable-next-line consistent-return
  size: (props, propName, componentName) => {
    if (props.size && !includes(sizes, props.size)) {
      return new Error(
        `Invalid prop size='${props.size}' supplied to ${componentName}`
      );
    }

    if (props.size && some(sizes, size => has(props, size))) {
      return new Error(
        `Seems that you've accidentially supplied boolean size along with size='${
          props.size
        }' to ${componentName}, please remove one of them. Otherwise boolean prop will overwrite the 'size' prop.`
      );
    }
  },
  ...getBooleanSizePropTypes()
};

const parseBooleanSize = props => {
  const sizeProps = {};

  forEach(sizes, size => {
    if (props[size]) {
      sizeProps.size = size;
    }
  });

  return sizeProps;
};

const withTextProps = OriginalComponent => {
  const DecoratedComponent = props => {
    const sizeProp = parseBooleanSize(props);

    const newProps = {
      ...omit(props, sizes),
      ...sizeProp
    };

    return <OriginalComponent {...newProps} />;
  };

  DecoratedComponent.propTypes = SizePropTypes;

  DecoratedComponent.displayName = OriginalComponent.displayName;

  return DecoratedComponent;
};

export default withTextProps;
