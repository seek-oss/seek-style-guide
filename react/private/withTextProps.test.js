import { shallow } from 'enzyme';
import React from 'react';

import withTextProps, { sizes } from './withTextProps';

const OriginalComponent = props => <div {...props} />;
const DecoratedComponent = withTextProps(OriginalComponent);

describe('withTextProps:', () => {
  let spy;

  beforeEach(() => {
    spy = jest.spyOn(global.console, 'error');
  });

  afterEach(() => {
    spy.mockRestore();
  });

  it('should pass down non-size props', () => {
    expect(
      shallow(<DecoratedComponent className="foo" bar baz={{ a: 1 }} />)
    ).toMatchSnapshot();
  });

  it('should throw prop validation error if size is unknown', () => {
    shallow(<DecoratedComponent size="head" />);

    const expectedError = expect.stringMatching(
      /Invalid prop size='head' supplied to OriginalComponent/
    );
    expect(spy).toBeCalledWith(expectedError);
  });

  it('should omit size prop if boolean size is supplied and throw prop validation error', () => {
    expect(
      shallow(<DecoratedComponent size="heading" hero />)
    ).toMatchSnapshot();

    const expectedError = expect.stringMatching(
      /Seems that you've accidentially supplied boolean size along with size='heading' to OriginalComponent, please remove one of them. Otherwise boolean prop will overwrite the 'size' prop/
    );
    expect(spy).toBeCalledWith(expectedError);
  });

  sizes.forEach(size => {
    describe(`*${size}*:`, () => {
      it('should pass down string size', () => {
        expect(shallow(<DecoratedComponent size={size} />)).toMatchSnapshot();
      });

      it('should convert boolean size into string size', () => {
        expect(
          shallow(<DecoratedComponent {...{ [size]: true }} />)
        ).toMatchSnapshot();
      });
    });
  });
});
