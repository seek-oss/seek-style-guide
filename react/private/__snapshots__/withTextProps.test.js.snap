// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withTextProps: *heading*: should convert boolean size into string size 1`] = `
<OriginalComponent
  size="heading"
/>
`;

exports[`withTextProps: *heading*: should pass down string size 1`] = `
<OriginalComponent
  size="heading"
/>
`;

exports[`withTextProps: *headline*: should convert boolean size into string size 1`] = `
<OriginalComponent
  size="headline"
/>
`;

exports[`withTextProps: *headline*: should pass down string size 1`] = `
<OriginalComponent
  size="headline"
/>
`;

exports[`withTextProps: *hero*: should convert boolean size into string size 1`] = `
<OriginalComponent
  size="hero"
/>
`;

exports[`withTextProps: *hero*: should pass down string size 1`] = `
<OriginalComponent
  size="hero"
/>
`;

exports[`withTextProps: *large*: should convert boolean size into string size 1`] = `
<OriginalComponent
  size="large"
/>
`;

exports[`withTextProps: *large*: should pass down string size 1`] = `
<OriginalComponent
  size="large"
/>
`;

exports[`withTextProps: *small*: should convert boolean size into string size 1`] = `
<OriginalComponent
  size="small"
/>
`;

exports[`withTextProps: *small*: should pass down string size 1`] = `
<OriginalComponent
  size="small"
/>
`;

exports[`withTextProps: *standard*: should convert boolean size into string size 1`] = `
<OriginalComponent
  size="standard"
/>
`;

exports[`withTextProps: *standard*: should pass down string size 1`] = `
<OriginalComponent
  size="standard"
/>
`;

exports[`withTextProps: *subheading*: should convert boolean size into string size 1`] = `
<OriginalComponent
  size="subheading"
/>
`;

exports[`withTextProps: *subheading*: should pass down string size 1`] = `
<OriginalComponent
  size="subheading"
/>
`;

exports[`withTextProps: should omit size prop if boolean size is supplied and throw prop validation error 1`] = `
<OriginalComponent
  size="hero"
/>
`;

exports[`withTextProps: should pass down non-size props 1`] = `
<OriginalComponent
  bar={true}
  baz={
    Object {
      "a": 1,
    }
  }
  className="foo"
/>
`;
